staging/src/k8s.io/client-go/testing/fixture.go

```diff
@@ -352,14 +390,96 @@ func (t *tracker) Add(obj runtime.Object) error {
 	return nil
 }
 
+func (t *tracker) trackFields(obj, newObj runtime.Object, manager string) error {
+	if t.fieldManager == nil {
+		// Do nothing if we don't have a field manager.
+		return nil
+	}
+
+	// Get the version off of the newObj
+	updObj, err := t.fieldManager.Update(obj, newObj, manager)
+	if err != nil {
+		return err
+	}
+
+	return setObject(obj, updObj.(*unstructured.Unstructured))
+	// transform the object
+}
+
 func (t *tracker) Create(gvr schema.GroupVersionResource, obj runtime.Object, ns string) error {
+	emptyObj := obj.DeepCopyObject()
+
+	// Empty the copy of the object.
+	value := reflect.ValueOf(emptyObj)
+	value.Elem().Set(reflect.New(value.Type().Elem()).Elem())
+
+	if err := t.trackFields(emptyObj, obj, "test-manager"); err != nil {
+		return err
+	}
+
 	return t.add(gvr, obj, ns, false)
 }
 
 func (t *tracker) Update(gvr schema.GroupVersionResource, obj runtime.Object, ns string) error {
+	objMeta, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+	orig, err := t.Get(gvr, ns, objMeta.GetName())
+	if err != nil {
+		return err
+	}
+	if err := t.trackFields(orig, obj, "test-manager"); err != nil {
+		return err
+	}
+
 	return t.add(gvr, obj, ns, true)
 }
 
+func (t *tracker) Apply(gvr schema.GroupVersionResource, obj, patch runtime.Object, ns string, manager string, force bool) error {
+	if t.fieldManager == nil {
+		panic("FieldManager should be installed in test to use Apply")
+	}
+
+	// Whether we should create the new object or not.
+	replace := obj != nil
+
+	if obj == nil {
+		// Create an empty object as the base object.
+		obj = patch.DeepCopyObject()
+		emptyObject(obj)
+		obj.GetObjectKind().SetGroupVersionKind(patch.GetObjectKind().GroupVersionKind())
+	}
+
+	newObj, err := t.fieldManager.Apply(obj, patch, manager, force)
+	if err != nil {
+		return err
+	}
+
+	// We need to get back the same underlying type of object as was
+	// given in, so we need to put the result back into obj. Reset
+	// it first.
+	if err := setObject(obj, newObj.(*unstructured.Unstructured)); err != nil {
+		return err
+	}
+
+	fmt.Printf("Final object after apply: %v\n", obj)
+
+	return t.add(gvr, obj, ns, replace)
```

Comment:
Do we need to do diff of old vs new to avoid NOP Apply() calls triggering updates?